### PR TITLE
GH-1515 moved store-related tests out of compliance

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #1515  

Briefly describe the changes proposed in this PR:

* converting various "compliance" tests into unit tests
* removed rdf4j-store-testsuite and rdf4j-store-compliance

The idea is to make our test rig simpler and more easily ran incrementally.

### CODE REVIEW

File: core/sail/api/src/test/java/org/eclipse/rdf4j/sail/helpers/NotifyingSailConnectionWrapperTest.java [hunk 1]
```diff
@@ -85,24 +82,8 @@ public void tearDown() throws SailException {
...
 		wrapper.addConnectionListener(listener);
-		addStatement("a");
-		assertThat(listener.getCount()).isEqualTo(1);
-		removeStatement("a");
-		assertThat(listener.getCount()).isEqualTo(0);
+		verify(delegate).addConnectionListener(listener);
 		wrapper.removeConnectionListener(listener);
-		addStatement("b");
-		assertThat(listener.getCount()).isEqualTo(0);
-	}
```
Comment (hmottestad) [inline core/sail/api/src/test/java/org/eclipse/rdf4j/sail/helpers/NotifyingSailConnectionWrapperTest.java]: Do some of the other tests check that you can add statements after the listener has been removed?

File: core/sail/base/pom.xml [hunk 2]
```diff
@@ -51,9 +51,49 @@
...
 	<build>
 		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<executions>
+					<execution>
+						<goals>
+							<goal>test-jar</goal>
+						</goals>
```
Comment (hmottestad) [inline core/sail/base/pom.xml]: Do we need a test jar?

Comment (hmottestad) [reply core/sail/base/pom.xml]: does this approach mean we need to build twice? Or that eclipse/IntelliJ won’t be able to build properly?

Comment (jeenbroekstra) [reply core/sail/base/pom.xml]: We need a test jar because normally maven modules don't include the test code. There are tests which are _defined_ at a certain abstract level but which need to executed in concrete implemenations. To be able to inherit from a common set of  tests, we use these test jars. This basically replaces the testsuite module we had for this purpose, and make it easier to just execute these tests as part of the module itself, instead of in a separate compliance module.

And no, it doesn't mean we need to build twice, I've been careful (hopefully) not to introduce any cyclic dependencies. This is a pretty common maven approach, it builds fine in Eclipse and though I haven't tested in IntelliJ I'd be very surprised if it had trouble with it.

File: core/sail/api/src/test/java/org/eclipse/rdf4j/sail/helpers/NotifyingSailConnectionWrapperTest.java (hunk 1 above)
Comment (jeenbroekstra) [reply core/sail/api/src/test/java/org/eclipse/rdf4j/sail/helpers/NotifyingSailConnectionWrapperTest.java]: I honestly don't know. I thought the whole test was badly written and simplified it so that I could at least integrate it. The point of this particular test is to check that connection listeners are properly added/remved, not to check if  statements can be added. It might be good to add more tests to this class to check for those things - part of reason I'm restructuring all this is also just to make it easier to find the right place to add a test.

Comment (hmottestad) [reply core/sail/api/src/test/java/org/eclipse/rdf4j/sail/helpers/NotifyingSailConnectionWrapperTest.java]: I have a bunch of tests for the notifying sail in the SHACL module. When I developed it I realized that there were a few edge cases that the notifying sail didn’t notify correctly. Things like adding and removing and re-adding the same statement in a single transaction.

**Comment by jeenbroekstra** [comment]
@rdf4j-bot run tests

**Comment by jeenbroekstra** [comment]
@rdf4j-bot run tests
